### PR TITLE
(PE-36952) ensure infra-serial is up to date

### PR DESCRIPTION
**Pull request by jonathannewman**
Please see individual commit comments.

### CODE REVIEW

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 1]
```diff
@@ -523,8 +523,6 @@
             {}
             auth-exts)))
 
-(defn seq-contains? [coll target] (some #(= target %) coll))
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:526]: This wasn't used, so I removed it.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 2]
```diff
@@ -658,15 +656,64 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Inventory File
 
+(def inventory-date-formatter
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:659]: Making the formatter once and reusing it saves some time and memory churn.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 3]
```diff
@@ -658,15 +656,64 @@
...
     (time-coerce/from-date date-time)))
 
+(schema/defn parse-date-time :- DateTime
+  "parses a date-time string into a DateTime instance"
+  [date-time :- schema/Str]
+  (time-format/parse inventory-date-formatter date-time))
+
 (def buffer-copy-size (* 64 1024))
 
+(schema/defn read-infra-nodes
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:676]: This was moved up from below.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 4]
```diff
@@ -818,41 +885,49 @@
...
+
+(schema/defn write-infra-serials-to-writer
+  [writer :- BufferedWriter
+   settings :- CaSettings]
+  (let [infra-serials (extract-active-infra-serials settings)]
+    (doseq [infra-serial infra-serials]
+      (.write writer (str infra-serial))
+      (.newLine writer))))
+
+(schema/defn generate-infra-serials!
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:924]: Renamed this by adding the `!` suffix to indicate that it makes changes.

File: test/unit/puppetlabs/services/ca/ca_testutils.clj [hunk 5]
```diff
@@ -73,7 +73,7 @@
    :serial                           (str cadir "/serial")
    :ruby-load-path                   jruby-testutils/ruby-load-path
    :gem-path                         jruby-testutils/gem-path
-   :infra-nodes-path                 (str cadir "/ca/infra_inventory.txt")
+   :infra-nodes-path                 (str cadir "/infra_inventory.txt")
```
Comment (jonathannewman) [inline test/unit/puppetlabs/services/ca/ca_testutils.clj:76]: This was a typo.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 6]
```diff
@@ -682,31 +729,33 @@
...
                                 (with-open [^BufferedReader reader (io/reader cert-inventory)]
                                   ;; copy all the existing content
                                   (loop [read-length (.read reader copy-buffer)]
-                                    (when (< 0 read-length)
-                                        (when (pos? read-length)
-                                          (.write writer (.array copy-buffer) 0 read-length))
-                                        (.clear copy-buffer)
-                                        (recur (.read reader copy-buffer)))))
+                                    ;; theoretically read can return 0, which means try again
+                                    (when (<= 0 read-length)
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:754]: This was incorrectly handling the `0` result case.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 7]
```diff
@@ -726,6 +776,13 @@
    (if (some? row-subject)
      (= (subs row-subject 1) cn-subject)
      false))
+(schema/defn is-not-expired? :- schema/Bool
```
Comment (justinstoller) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:779]: Needs a newline separating this function from the preceding one.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 8]
```diff
@@ -717,7 +766,8 @@
                             (log/trace (i18n/trs "Finish append to inventory file. ")))]
     (common/with-safe-write-lock inventory-lock inventory-lock-descriptor inventory-lock-timeout-seconds
       (log/debug (i18n/trs "Append \"{1}\" to inventory file {0}" cert-inventory entry))
-      (ks-file/atomic-write cert-inventory stream-content-fn))))
+      (ks-file/atomic-write cert-inventory stream-content-fn))
+    (maybe-write-to-infra-serial! serial-number cert-name settings)))
```
Comment (justinstoller) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:770]: Shouldn't we put this in the inventory-lock above?

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 4 above)
Comment (justinstoller) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:924]: Just a sanity check here:
When a cert is renewed the new serial will be appended to the infra-serials file, but when the CA restarts, reloads, or the infra node certname file is updated the whole infra-serials file will be rewritten with the latest serial for each infra node. That's the intended behavior?

**Review by justinstoller — COMMENTED** [review]
Looks good overall.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 8 above)
Comment (jonathannewman) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:770]: The serial isn't locked under any other circumstances, but it certainly wouldn't hurt.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 4 above)
Comment (jonathannewman) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:924]: In the startup case, with these changes to `generate-infra-serials!` we are now looking in the inventory as well for matching serial numbers and adding them.  So at startup there should be technically no work to do, unless someone manually changed the files.  The idea is that the infra-serial should always be more correct now

Comment (jonathannewman) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:924]: So to answer your question, the intent is that we append the infra-serial on demand when a new one is generated, and when we restart, we update the state based on the current state on the filesystem, including all known valid serial numbers for the infra-nodes, not just the most recent one.  This helps at revocation time to ensure the infra-crl is updated with all possible certs for the host being revoked.

**Comment by jonathannewman** [comment]
Updated to address the two issues mentioned.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 4 above)
Comment (justinstoller) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:925]: Thanks for that.

**Review by justinstoller — APPROVED** [review]

**Pull request merged**